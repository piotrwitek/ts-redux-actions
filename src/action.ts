import {
  StringType,
  EmptyAction,
  PayloadAction,
  PayloadMetaAction,
  MetaAction,
} from './type-helpers';
import {
  checkIsEmpty,
  throwIsEmpty,
  checkInvalidActionType,
  throwInvalidActionCreator,
} from './utils/validation';

export function action<T extends StringType, M>(
  type: T,
  payload: undefined,
  meta: M
): { type: T; meta: M };

export function action<T extends StringType, P, M>(
  type: T,
  payload: P,
  meta: M
): { type: T; payload: P; meta: M };

export function action<T extends StringType, P>(
  type: T,
  payload: P
): { type: T; payload: P };

export function action<T extends StringType>(type: T): { type: T };

/**
 * @description flux standard action factory
 * @example
 * ```
 * const add = (amount: number, meta?: MetaShape) => action('INCREMENT', amount, meta);
 * ```
 */
export function action<T extends StringType, P = undefined, M = undefined>(
  type: T,
  payload?: P,
  meta?: M
) {
  if (checkIsEmpty(type)) {
    throwIsEmpty(1);
  }

  if (checkInvalidActionType(type)) {
    throwInvalidActionCreator(1);
  }

  return { type, payload, meta } as any;
}
