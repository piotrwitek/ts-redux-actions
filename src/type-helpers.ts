/**
 * PUBLIC API
 */

/**
 * @desc Type representing Type Constant
 */
export type TypeConstant = string;

/**
 * @desc Type representing Generic Action
 */
export type Action<TType extends TypeConstant = TypeConstant> = {
  type: TType;
};

/**
 * @desc Type representing Generic ActionCreator
 */
export type ActionCreator<TType extends TypeConstant> = (
  ...args: any[]
) => Action<TType>;

/**
 * @desc Type representing Generic Reducer
 */
export type Reducer<TState, TAction extends Action> = (
  state: TState | undefined,
  action: TAction
) => TState;

/**
 * @desc Action without Payload
 */
export type EmptyAction<TType extends TypeConstant> = {
  type: TType;
};

/**
 * @desc Action with only Payload
 */
export type PayloadAction<TType extends TypeConstant, TPayload> = {
  type: TType;
  payload: TPayload;
};

/**
 * @desc Action with only Meta
 */
export type MetaAction<TType extends TypeConstant, TMeta> = {
  type: TType;
  meta: TMeta;
};

/**
 * @desc Action with both Payload and Meta
 */
export type PayloadMetaAction<TType extends TypeConstant, TPayload, TMeta> = {
  type: TType;
  payload: TPayload;
  meta: TMeta;
};

/**
 * @desc Action Creator producing EmptyAction
 */
export type EmptyAC<TType extends TypeConstant> = () => EmptyAction<TType>;

/**
 * @desc Action Creator producing PayloadAction
 */
export type PayloadAC<TType extends TypeConstant, TPayload> = (
  payload: TPayload
) => PayloadAction<TType, TPayload>;

/**
 * @desc Action Creator producing PayloadMetaAction
 */
export type PayloadMetaAC<TType extends TypeConstant, TPayload, TMeta> = (
  payload: TPayload,
  meta: TMeta
) => PayloadMetaAction<TType, TPayload, TMeta>;

/**
 * @desc Type representing type getter on Action Creator instance
 */
export interface TypeMeta<TType extends TypeConstant> {
  getType?: () => TType;
}

/**
 * @desc Infers Action union-type from action-creator map object
 */
export type ActionType<
  TActionCreatorOrMap extends any
> = TActionCreatorOrMap extends ActionCreator<TypeConstant>
  ? ReturnType<TActionCreatorOrMap>
  : TActionCreatorOrMap extends Record<any, any>
  ? {
      [K in keyof TActionCreatorOrMap]: ActionType<TActionCreatorOrMap[K]>
    }[keyof TActionCreatorOrMap]
  : TActionCreatorOrMap extends infer R // TODO: should be just never but compiler yell with circularly references itself error
  ? never
  : never;

/**
 * @desc Infers State object from reducer map object
 */
export type StateType<
  TReducerOrMap extends any
> = TReducerOrMap extends Reducer<any, any>
  ? ReturnType<TReducerOrMap>
  : TReducerOrMap extends Record<any, any>
  ? { [K in keyof TReducerOrMap]: StateType<TReducerOrMap[K]> }
  : never;

/**
 * INTERNAL API
 */

/** @private */
export type ActionBuilderConstructor<
  TType extends TypeConstant,
  TPayload extends any = undefined,
  TMeta extends any = undefined
> = [TMeta] extends [undefined]
  ? [TPayload] extends [undefined]
    ? unknown extends TPayload
      ? PayloadAC<TType, TPayload>
      : unknown extends TMeta
      ? PayloadMetaAC<TType, TPayload, TMeta>
      : EmptyAC<TType>
    : PayloadAC<TType, TPayload>
  : PayloadMetaAC<TType, TPayload, TMeta>;

/** @private */
export type ActionBuilderMap<
  TType extends TypeConstant,
  TActionProps extends any,
  TPayloadArg extends any = undefined,
  TMetaArg extends any = undefined
> = [TMetaArg] extends [undefined]
  ? [TPayloadArg] extends [undefined]
    ? () => { type: TType } & TActionProps
    : (payload: TPayloadArg) => { type: TType } & TActionProps
  : (payload: TPayloadArg, meta: TMetaArg) => { type: TType } & TActionProps;
