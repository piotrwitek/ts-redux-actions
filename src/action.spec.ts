import * as T from './type-helpers'; // type-tests global
import { action } from './action';

it.skip('skip', () => undefined);

describe('with symbol', () => {
  const ACTION = Symbol(1);
  const withSymbol = () => action(ACTION as any);
  // @dts-jest:pass:snap
  withSymbol(); // => { type: ACTION }
});

describe('with type only', () => {
  const withTypeOnly = () => action('ACTION');
  // @dts-jest:pass:snap
  withTypeOnly(); // => { type: 'ACTION' }
});

describe('with payload', () => {
  const withPayload = (message: string) => action('ACTION', message);
  // @dts-jest:pass:snap
  withPayload('Hello!'); // => { type: 'ACTION',payload: 'Hello!' }
});

describe('with optional payload', () => {
  const withOptionalPayload = (message?: string) => action('ACTION', message);
  // @dts-jest:pass:snap
  withOptionalPayload(); // => { type: 'ACTION' }
  // @dts-jest:pass:snap
  withOptionalPayload('Hello!'); // => { type: 'ACTION', payload: 'Hello!' }
});

describe('with meta', () => {
  const withMeta = (scope: string) => action('ACTION', undefined, scope);
  // @dts-jest:pass:snap
  withMeta('info'); // => { type: 'ACTION', meta: 'info' }
});

describe('with optional meta', () => {
  const withOptionalMeta = (scope?: string) =>
    action('ACTION', undefined, scope);
  // @dts-jest:pass:snap
  withOptionalMeta(); // => { type: 'ACTION' }
  // @dts-jest:pass:snap
  withOptionalMeta('info'); // => { type: 'ACTION', meta: 'info' }
});

describe('with payload and meta', () => {
  const withPayloadAndMeta = (message: string, scope: string) =>
    action('ACTION', message, scope);
  // @dts-jest:pass:snap
  withPayloadAndMeta('Hello!', 'info'); // => { type: 'ACTION', payload: 'Hello!', meta: 'info' }
});

describe('with optional payload and meta', () => {
  const withOptionalPayloadAndMeta = (scope: string, message?: string) =>
    action('ACTION', message, scope);
  // @dts-jest:pass:snap
  withOptionalPayloadAndMeta('info'); // => { type: 'ACTION', meta: 'info' }
  // @dts-jest:pass:snap
  withOptionalPayloadAndMeta('info', 'Hello!'); // => { type: 'ACTION', payload: 'Hello!', meta: 'info' }
});

describe('with payload and optional meta', () => {
  const withPayloadAndOptionalMeta = (message: string, scope?: string) =>
    action('ACTION', message, scope);
  // @dts-jest:pass:snap
  withPayloadAndOptionalMeta('Hello!'); // => { type: 'ACTION', payload: 'Hello!' }
  // @dts-jest:pass:snap
  withPayloadAndOptionalMeta('Hello!', 'info'); // => { type: 'ACTION', payload: 'Hello!', meta: 'info' }
});

describe('with optional payload and optional meta', () => {
  const withOptionalPayloadAndOptionalMeta = (
    message?: string,
    scope?: string
  ) => action('ACTION', message, scope);
  // @dts-jest:pass:snap
  withOptionalPayloadAndOptionalMeta(); // => { type: 'ACTION' }
  // @dts-jest:pass:snap
  withOptionalPayloadAndOptionalMeta('Hello!'); // => { type: 'ACTION', payload: 'Hello!' }
  // @dts-jest:pass:snap
  withOptionalPayloadAndOptionalMeta('Hello!', 'info'); // => { type: 'ACTION', payload: 'Hello!', meta: 'info' }
});

describe('with error', () => {
  const withError = (error: boolean) =>
    action('ACTION', undefined, undefined, error);
  // @dts-jest:pass:snap
  withError(true); // => { type: 'ACTION', error: true }
});

describe('with optional error', () => {
  const withOptionalError = (error?: boolean) =>
    action('ACTION', undefined, undefined, error);
  // @dts-jest:pass:snap
  withOptionalError(); // => { type: 'ACTION' }
  // @dts-jest:pass:snap
  withOptionalError(true); // => { type: 'ACTION', error: true }
});

describe('with payload, meta, and error', () => {
  const withPayloadMetaAndError = (
    message: string,
    scope: string,
    error: boolean
  ) => action('ACTION', message, scope, error);
  // @dts-jest:pass:snap
  withPayloadMetaAndError('Hello!', 'info', true); // => { type: 'ACTION', payload: 'Hello!', meta: 'info', error: true }
});

describe('with payload, meta, and optional error', () => {
  const withPayloadMetaAndOptionalError = (
    message: string,
    scope: string,
    error?: boolean
  ) => action('ACTION', message, scope, error);
  // @dts-jest:pass:snap
  withPayloadMetaAndOptionalError('Hello!', 'info'); // => { type: 'ACTION', payload: 'Hello!', meta: 'info' }
  // @dts-jest:pass:snap
  withPayloadMetaAndOptionalError('Hello!', 'info', true); // => { type: 'ACTION', payload: 'Hello!', meta: 'info', error: true }
});

describe('with optional payload, optional meta, and optional error', () => {
  const withOptionalPayloadOptionalMetaAndOptionalError = (
    message?: string,
    scope?: string,
    error?: boolean
  ) => action('ACTION', message, scope, error);
  // @dts-jest:pass:snap
  withOptionalPayloadOptionalMetaAndOptionalError(); // => { type: 'ACTION' }
  // @dts-jest:pass:snap
  withOptionalPayloadOptionalMetaAndOptionalError('Hello!'); // => { type: 'ACTION', payload: 'Hello!' }
  // @dts-jest:pass:snap
  withOptionalPayloadOptionalMetaAndOptionalError('Hello!', 'info'); // => { type: 'ACTION', payload: 'Hello!', meta: 'info' }
  // @dts-jest:pass:snap
  withOptionalPayloadOptionalMetaAndOptionalError('Hello!', 'info', true); // => { type: 'ACTION', payload: 'Hello!', meta: 'info', error: true }
});
