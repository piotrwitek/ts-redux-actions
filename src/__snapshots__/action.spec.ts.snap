// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withError(true) 1`] = `"{ type: \\"ACTION\\"; error: boolean; }"`;

exports[`withMeta('info') 1`] = `"{ type: \\"ACTION\\"; meta: string; }"`;

exports[`withOptionalError() 1`] = `"{ type: \\"ACTION\\"; error: boolean | undefined; }"`;

exports[`withOptionalError(true) 1`] = `"{ type: \\"ACTION\\"; error: boolean | undefined; }"`;

exports[`withOptionalMeta('info') 1`] = `"{ type: \\"ACTION\\"; meta: string | undefined; }"`;

exports[`withOptionalMeta() 1`] = `"{ type: \\"ACTION\\"; meta: string | undefined; }"`;

exports[`withOptionalPayload('Hello!') 1`] = `"{ type: \\"ACTION\\"; payload: string | undefined; }"`;

exports[`withOptionalPayload() 1`] = `"{ type: \\"ACTION\\"; payload: string | undefined; }"`;

exports[`withOptionalPayloadAndMeta('info') 1`] = `"{ type: \\"ACTION\\"; payload: string | undefined; meta: string; }"`;

exports[`withOptionalPayloadAndMeta('info', 'Hello!') 1`] = `"{ type: \\"ACTION\\"; payload: string | undefined; meta: string; }"`;

exports[`withOptionalPayloadAndOptionalMeta('Hello!') 1`] = `"{ type: \\"ACTION\\"; payload: string | undefined; meta: string | undefined; }"`;

exports[`withOptionalPayloadAndOptionalMeta('Hello!', 'info') 1`] = `"{ type: \\"ACTION\\"; payload: string | undefined; meta: string | undefined; }"`;

exports[`withOptionalPayloadAndOptionalMeta() 1`] = `"{ type: \\"ACTION\\"; payload: string | undefined; meta: string | undefined; }"`;

exports[`withOptionalPayloadOptionalMetaAndOptionalError('Hello!') 1`] = `"{ type: \\"ACTION\\"; payload: string | undefined; meta: string | undefined; error: boolean | undefined; }"`;

exports[`withOptionalPayloadOptionalMetaAndOptionalError('Hello!', 'info') 1`] = `"{ type: \\"ACTION\\"; payload: string | undefined; meta: string | undefined; error: boolean | undefined; }"`;

exports[`withOptionalPayloadOptionalMetaAndOptionalError('Hello!', 'info', true) 1`] = `"{ type: \\"ACTION\\"; payload: string | undefined; meta: string | undefined; error: boolean | undefined; }"`;

exports[`withOptionalPayloadOptionalMetaAndOptionalError() 1`] = `"{ type: \\"ACTION\\"; payload: string | undefined; meta: string | undefined; error: boolean | undefined; }"`;

exports[`withPayload('Hello!') 1`] = `"{ type: \\"ACTION\\"; payload: string; }"`;

exports[`withPayloadAndMeta('Hello!', 'info') 1`] = `"{ type: \\"ACTION\\"; payload: string; meta: string; }"`;

exports[`withPayloadAndOptionalMeta('Hello!') 1`] = `"{ type: \\"ACTION\\"; payload: string; meta: string | undefined; }"`;

exports[`withPayloadAndOptionalMeta('Hello!', 'info') 1`] = `"{ type: \\"ACTION\\"; payload: string; meta: string | undefined; }"`;

exports[`withPayloadMetaAndError('Hello!', 'info', true) 1`] = `"{ type: \\"ACTION\\"; payload: string; meta: string; error: boolean; }"`;

exports[`withPayloadMetaAndOptionalError('Hello!', 'info') 1`] = `"{ type: \\"ACTION\\"; payload: string; meta: string; error: boolean | undefined; }"`;

exports[`withPayloadMetaAndOptionalError('Hello!', 'info', true) 1`] = `"{ type: \\"ACTION\\"; payload: string; meta: string; error: boolean | undefined; }"`;

exports[`withSymbol() 1`] = `"{ type: any; }"`;

exports[`withTypeOnly() 1`] = `"{ type: \\"ACTION\\"; }"`;
